### PR TITLE
Add `date_trunc()` simplification rules

### PR DESCRIPTION
**Pull request by rcurtin**
Hello there,

This is my first non-trivial PR to DuckDB.  I hope that I have followed the procedure correctly ... please tell me if not and I can fix things. :)

I noticed that for a query of the form `select <things> from <somewhere> where date_trunc(part, some_column) > 'some constant date')` would keep the `date_trunc()` in the filters that were pushed down.  However, if you have a `date_trunc()` on one side and a constant on the other side, you can actually do a rewrite here to get rid of the `date_trunc()` by adjusting the constant accordingly.

As a really simple example, these two are equivalent:

```
select count(*) from lineitem where date_trunc('month', l_shipdate) >= '1997-01-01';
select count(*) from lineitem where l_shipdate >= '1997-01-01';
```

Sometimes the situation is a little more complex:

```
select count(*) from lineitem where date_trunc('month', l_shipdate) > '1997-01-01';
select count(*) from lineitem where l_shipdate >= '1997-01-01';
```

In any case, whenever you have `date_trunc(part, column) <comparison_op> constant_rhs`, with a little algebraic rewrite you can produce `column <possibly_new_op> possibly_new_constant_rhs`.  This in turn allows these filters to be pushed down directly into the table.  For instance, before this PR, here is a query plan for the queries above:

```
D explain select count(*) from lineitem where date_trunc('month', l_shipdate) > '1997-01-01';

┌─────────────────────────────┐
│┌───────────────────────────┐│
││       Physical Plan       ││
│└───────────────────────────┘│
└─────────────────────────────┘
┌───────────────────────────┐
│    UNGROUPED_AGGREGATE    │
│    ────────────────────   │
│        Aggregates:        │
│        count_star()       │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│           FILTER          │
│    ────────────────────   │
│    (date_trunc('month',   │
│  l_shipdate) > '1997-01-01│
│          '::DATE)         │
│                           │
│       ~11997210 Rows      │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│         SEQ_SCAN          │
│    ────────────────────   │
│      Table: lineitem      │
│   Type: Sequential Scan   │
│                           │
│        Projections:       │
│         l_shipdate        │
│                           │
│       ~59986052 Rows      │
└───────────────────────────┘
```

And afterwards, much simpler:

```
D explain select count(*) from lineitem where date_trunc('month', l_shipdate) > '1997-01-01';

┌─────────────────────────────┐
│┌───────────────────────────┐│
││       Physical Plan       ││
│└───────────────────────────┘│
└─────────────────────────────┘
┌───────────────────────────┐
│    UNGROUPED_AGGREGATE    │
│    ────────────────────   │
│        Aggregates:        │
│        count_star()       │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│         SEQ_SCAN          │
│    ────────────────────   │
│      Table: lineitem      │
│   Type: Sequential Scan   │
│                           │
│          Filters:         │
│ l_shipdate>='1997-02-01': │
│           :DATE           │
│                           │
│       ~11997210 Rows      │
└───────────────────────────┘
```

Timing-wise, for these same queries:

 * Before: `Run Time (s): real 0.166 user 3.729545 sys 0.127580`
 * After: `Run Time (s): real 0.036 user 0.545248 sys 0.142857`

So, a significant speedup for this (best-case) query.  To work around any caching effects, I ran those in a completely new DuckDB session where I loaded a pre-built TPCH dataset with scale factor 10 that I had computed earlier.

I wrote test cases to ensure the correctness of the optimization and also that the optimization applies.

As I mentioned I have never contributed to DuckDB before although I have certainly been deeply involved with the development of other database query optimizers, so I know that there are going to be assumptions that the query optimizer has that I don't currently have the context to know.  It is possible that I have violated these in places... just let me know if so and I will get it fixed (but it seems to work fine for me in my test cases at least!).

### CODE REVIEW

**Comment by hawkfish** [comment]
Hi @rcurtin - thanks for the PR!  

Before I dig in any further, can you fix the formatting (`make format-head`) and look at the Windows failure?  Also, it would be good to check this out with time zones (e.g., tests across DST boundaries and incomplete hour offsets?

**Comment by rcurtin** [comment]
Working my way through the CI failures, I should have it all working shortly.  Looks like I just missed some includes where gcc doesn't have a problem but MSVC (rightly) does.  It seems like the CI only runs if I select 'ready for review'---if I misunderstood that let me know.

Can you give me an example of an incomplete hour offset?  I believe everything should work in these cases (as well as the DST case) but I agree it would be good to add more test cases for those.

File: test/optimizer/date_trunc_simplification.test [hunk 1]
```diff
@@ -2,8 +2,10 @@
 # description: test DATE_TRUNC() constant simplifications
 # group: [optimizer]
 
+require icu
```
Comment (hawkfish) [inline test/optimizer/date_trunc_simplification.test]: Small thing - we generally put the ICU tests in a separate file. This allows the feature to be tested in an environment where ICU is not available/loaded (e.g., WASM)

File: test/optimizer/date_trunc_simplification.test [hunk 2]
```diff
@@ -12,7 +14,7 @@ statement ok
 PRAGMA explain_output = OPTIMIZED_ONLY;
```
Comment (hawkfish) [inline test/optimizer/date_trunc_simplification.test:9]: An alternative to looking at explain output would be to use a loop that checks the results with and without the optimiser? This would make sure that all tests give the correct results instead of the one that are explicitly added.

File: src/optimizer/rule/date_trunc_simplification.cpp [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+				} else {
+					expr.SetExpressionTypeUnsafe(ExpressionType::COMPARE_LESSTHANOREQUALTO);
+				}
+			}
+
+			changes_made = true;
+			return nullptr;
+		}
+
+	default:
```
Comment (hawkfish) [inline src/optimizer/rule/date_trunc_simplification.cpp:284]: Can we add `IS [NOT] DISTINCT` cases as well?

**Review by hawkfish — CHANGES_REQUESTED** [review]
This is pretty great - thanks for doing it. I've been a bit busy recently, but I've made some basic comments. I'll try to do a deeper dive this week.

File: test/optimizer/date_trunc_simplification.test (hunk 1 above)
Comment (rcurtin) [reply test/optimizer/date_trunc_simplification.test]: I see!  Split into two files in bdc1a12760.  Let me know if I should name them differently.

File: test/optimizer/date_trunc_simplification.test (hunk 2 above)
Comment (rcurtin) [reply test/optimizer/date_trunc_simplification.test:9]: Correct me if I am wrong here: I believe that the `enable_verification` at the top of the test will test all the queries with and without the optimizer, so I think we are covered there.  The `explain analyze` test cases are specifically to ensure that the optimization is actually applying at all---it would be really easy for a bug in the heuristic to mean that it never applied the optimizations (and in that case, it always gives the same output as the unoptimized version, so the "failure" is not picked up).  I'm not 100% sure I'm answering what you were getting at, though.

File: src/optimizer/rule/date_trunc_simplification.cpp (hunk 3 above)
Comment (rcurtin) [reply src/optimizer/rule/date_trunc_simplification.cpp:284]: Ah!  I overlooked those.  After working through the algebra I implemented them (they're quite similar but not exactly the same as the `<>` and `=` cases).  I also added a bunch of tests to handle `NULL`s on the rhs or lhs in 56f49e0608.

File: test/optimizer/date_trunc_simplification.test (hunk 2 above)
Comment (hawkfish) [reply test/optimizer/date_trunc_simplification.test:9]: No you are correct - good point!

File: src/include/duckdb/optimizer/rule/date_trunc_simplification.hpp [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+namespace duckdb {
+
+// DateTruncSimplificationRule rewrites an expression of the form
+//
+//    date_trunc(part, column) <op> const_rhs
+//
+// such that the date_trunc is instead applied to the constant RHS and then simplified further.
+// The rules applied are as follows:
+//
+//   date_trunc(part, column) >= const_rhs   -->   column >= date_trunc(part, date_add(const_rhs, INTERVAL 1 part))
```
Comment (hawkfish) [inline src/include/duckdb/optimizer/rule/date_trunc_simplification.hpp:23]: I don't think this is correct.  For example, if I use TPC-H at SF 0.01, these two queries produce different row counts:

```sql
select count(*)
from lineitem
where date_trunc('day', l_shipdate) >= '1994-01-01';
-- 43454
select count(*)
from lineitem
where l_shipdate >= date_trunc('day', date_add('1994-01-01'::DATE, INTERVAL 1 day));
-- 43432
```

Comment (rcurtin) [reply src/include/duckdb/optimizer/rule/date_trunc_simplification.hpp:23]: The key is in the line of the comment just below:

```
//    - but if date_trunc(const_rhs) = const_rhs, then we can do column >= const_rhs
```

so, really the optimized version is not `l_shipdate >= date_trunc('day', date_add('1994-01-01'::DATE, INTERVAL 1 day))` but because `date_trunc('day', '1994-01-01'::DATE) = '1994-01-01::DATE'`, we can optimize instead to just `l_shipdate >= '1994-01-01'::DATE`.  This edge case caught me out at first.  It only applies to `>=` and `<`, not for `>` and `<=`.

With the optimization applied, here is the query plan:

```
D explain select count(*) from lineitem where date_trunc('day', l_shipdate) >= '1994-01-01';

┌─────────────────────────────┐
│┌───────────────────────────┐│
││       Physical Plan       ││
│└───────────────────────────┘│
└─────────────────────────────┘
┌───────────────────────────┐
│    UNGROUPED_AGGREGATE    │
│    ────────────────────   │
│        Aggregates:        │
│        count_star()       │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│         SEQ_SCAN          │
│    ────────────────────   │
│      Table: lineitem      │
│   Type: Sequential Scan   │
│                           │
│          Filters:         │
│ l_shipdate>='1994-01-01': │
│           :DATE           │
│                           │
│       ~1200243 Rows       │
└───────────────────────────┘
```

so it does properly catch that edge case.

Comment (hawkfish) [reply src/include/duckdb/optimizer/rule/date_trunc_simplification.hpp:23]: Thanks - I'll try to have another look this week. I really want this for 1.4, so I will get to it, but I spent all of last week trying to fix a Postgres incompatibility that is impossible to fix...

Comment (rcurtin) [reply src/include/duckdb/optimizer/rule/date_trunc_simplification.hpp:23]: No worries!  I also maintain a large open-source project and often spend entire weeks fixing things that are impossible to fix.  (Literally what I am doing right this moment.  Who knows if I will have success.  Probably not.  Sure feels like that.  Sigh.)  So anyway your pace is fine here, no rush from my end.

Comment (hawkfish) [reply src/include/duckdb/optimizer/rule/date_trunc_simplification.hpp:23]: This makes sense to me now: The constant just needs to be rounded up to the next bin boundary.

File: test/optimizer/date_trunc_simplification.test [hunk 5]
```diff
@@ -0,0 +1,312 @@
+# name: test/optimizer/date_trunc_simplification.test
+# description: test DATE_TRUNC() constant simplifications
+# group: [optimizer]
+
+statement ok
```
Comment (hawkfish) [inline test/optimizer/date_trunc_simplification.test:5]: Nit: We usually put these after the pragmas and other config statements.

File: src/include/duckdb/optimizer/rule/date_trunc_simplification.hpp [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+//   date_trunc(part, column) <  const_rhs   -->   column <  date_trunc(part, date_add(const_rhs, INTERVAL 1 part))
+//    - but if date_trunc(const_rhs) = const_rhs, then we can do column < const_rhs
+//
+//   date_trunc(part, column) == const_rhs   -->   column >= date_trunc(part, const_rhs) AND
+//                                                    column <  date_trunc(part, date_add(const_rhs, INTERVAL 1 part))
+//    - but if date_trunc(const_rhs) != const_rhs, then this one is unsatisfiable
+//
+//   date_trunc(part, column) <> const_rhs   -->   column <  date_trunc(part, const_rhs) OR
+//                                                    column >= date_trunc(part, date_add(const_rhs, INTERVAL 1 part))
+//    - but if date_trunc(const_rhs) != const_rhs, then this is always satisfied
```
Comment (hawkfish) [inline src/include/duckdb/optimizer/rule/date_trunc_simplification.hpp:40]: Can we have an explanation of the` IS [NOT] DISTINCT` rules?

File: test/optimizer/date_trunc_simplification_icu.test [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+# description: test DATE_TRUNC() constant simplifications that require icu
+# group: [optimizer]
+
+require icu
+
+statement ok
+PRAGMA enable_verification;
+
+#
+# check operation with hour offsets that aren't complete hours
```
Comment (hawkfish) [inline test/optimizer/date_trunc_simplification_icu.test:14]: Unfortunately, to test this, you need to use time zones that are not hour offsets, which is not the same as UTC and instants specified by offsets. The reason is that the UTC binning rules are what happens without ICU (we even optimise to use them if we are in Gregorian/UTC as they are faster). So we need to check that the rules are being correctly applied for different time zones.

Good ones to use are:

|        name         |     abbrev     | utc_offset | is_dst |
|---------------------|----------------|------------|-------:|
| America/Los_Angeles | PST            | -07:00:00  | true   |
| America/St_Johns    | CNT            | -02:30:00  | true   |
| Asia/Kathmandu      | Asia/Kathmandu | 05:45:00   | false  |
| Europe/Amsterdam    | CET            | 02:00:00   | true   |
| Pacific/Chatham     | NZ-CHAT        | 12:45:00   | false  |

You don't need to use _all_ of them, but one or two would be good.

File: test/optimizer/date_trunc_simplification_icu.test [hunk 8]
```diff
@@ -0,0 +1,84 @@
+# name: test/optimizer/date_trunc_simplification_icu.test
+# description: test DATE_TRUNC() constant simplifications that require icu
+# group: [optimizer]
+
+require icu
+
+statement ok
+PRAGMA enable_verification;
```
Comment (hawkfish) [inline test/optimizer/date_trunc_simplification_icu.test:8]: Heh you also need to set the calendar:

```sql
set Calendar='gregorian';
```
Otherwise if the testing locale's default calendar is not Gregorian, the tests will fail.  We set it for CI, but I've talked to people from China who had a different default calendar(!)

File: src/optimizer/rule/date_trunc_simplification.cpp [hunk 9]
```diff
@@ -0,0 +1,483 @@
...
+	root = std::move(op);
+}
+
+unique_ptr<Expression> DateTruncSimplificationRule::Apply(LogicalOperator &op, vector<reference<Expression>> &bindings,
+                                                          bool &changes_made, bool is_root) {
+	auto &expr = bindings[0].get().Cast<BoundComparisonExpression>();
+	auto comparison_type = expr.GetExpressionType();
+
+	auto &date_part = bindings[2].get().Cast<BoundConstantExpression>();
+	auto &column_part = bindings[3].get().Cast<BoundColumnRefExpression>();
```
Comment (hawkfish) [inline src/optimizer/rule/date_trunc_simplification.cpp:52]: What if it isn't a column reference but a nested expression?

File: src/optimizer/rule/date_trunc_simplification.cpp [hunk 10]
```diff
@@ -0,0 +1,483 @@
...
+	auto &rhs = bindings[4].get().Cast<BoundConstantExpression>();
+
+	// Determine whether or not the column name is on the lhs or rhs.
+	const bool col_is_lhs = (expr.left->GetExpressionClass() == ExpressionClass::BOUND_FUNCTION);
+
+	// We want to treat rhs >= col equivalently to col <= rhs.
+	// So, get the expression type if it was ordered such that the constant was actually on the right hand side.
+	ExpressionType rhs_comparison_type = comparison_type;
+	if (!col_is_lhs) {
+		switch (comparison_type) {
```
Comment (hawkfish) [inline src/optimizer/rule/date_trunc_simplification.cpp]: `FlipComparisonExpression`

File: src/optimizer/rule/date_trunc_simplification.cpp [hunk 11]
```diff
@@ -0,0 +1,483 @@
...
+		//  column < date_trunc(part, date_add(constant_rhs, INTERVAL 1 part)) AND
+		//  column IS NOT NULL)
+		//
+		{
+			if (rhs.value.IsNull()) {
+				// Return 'column IS NULL'.
+				auto op = make_uniq<BoundOperatorExpression>(ExpressionType::OPERATOR_IS_NULL, LogicalType::BOOLEAN);
+				op->children.push_back(column_part.Copy());
+				return std::move(op);
+			} else {
```
Comment (hawkfish) [inline src/optimizer/rule/date_trunc_simplification.cpp]: I'm wondering if we shouldn't fall through or share code with equality here?  Share the cases, but do the `NULL` check for IND.

File: src/optimizer/rule/date_trunc_simplification.cpp [hunk 12]
```diff
@@ -0,0 +1,483 @@
...
+				auto isnotnull =
+				    make_uniq<BoundOperatorExpression>(ExpressionType::OPERATOR_IS_NOT_NULL, LogicalType::BOOLEAN);
+				isnotnull->children.push_back(column_part.Copy());
+
+				return make_uniq<BoundConjunctionExpression>(ExpressionType::CONJUNCTION_AND, std::move(comp),
+				                                             std::move(isnotnull));
+			}
+		}
+
+	case ExpressionType::COMPARE_DISTINCT_FROM:
```
Comment (hawkfish) [inline src/optimizer/rule/date_trunc_simplification.cpp]: Share code with <>?

File: src/optimizer/rule/date_trunc_simplification.cpp [hunk 13]
```diff
@@ -0,0 +1,483 @@
...
+	default:
+		return nullptr;
+	}
+}
+
+string DateTruncSimplificationRule::DatePartToFunc(const DatePartSpecifier &date_part) {
+	switch (date_part) {
+	// These specifiers can be used as intervals.
+	case DatePartSpecifier::YEAR:
+		return "to_year";
```
Comment (hawkfish) [inline src/optimizer/rule/date_trunc_simplification.cpp]: `to_years`.  Maybe we should add some quick coverage tests for all these cases?

**Review by hawkfish — CHANGES_REQUESTED** [review]
Getting closer just a few more things.

File: test/optimizer/date_trunc_simplification.test (hunk 5 above)
Comment (rcurtin) [reply test/optimizer/date_trunc_simplification.test:5]: :+1: done in dafaeec211.

File: src/include/duckdb/optimizer/rule/date_trunc_simplification.hpp (hunk 6 above)
Comment (rcurtin) [reply src/include/duckdb/optimizer/rule/date_trunc_simplification.hpp:40]: Added in 41416bd67b.

File: test/optimizer/date_trunc_simplification_icu.test (hunk 7 above)
Comment (rcurtin) [reply test/optimizer/date_trunc_simplification_icu.test:14]: I don't understand here; I am not a timezone expert although I am capable of showing up for meetings on time.  Do you mean that instead of entering times as, e.g., `'2025-01-01 01:01:00+02:10'` I should instead use `'2025-01-01 01:01:00 PST'` and similar?  Would I also need to set a timezone as part of the test case so that the timezone of the test runner isn't used instead?  I'm not sure I understand what you're going for here, sorry about that.

File: test/optimizer/date_trunc_simplification_icu.test (hunk 8 above)
Comment (rcurtin) [reply test/optimizer/date_trunc_simplification_icu.test:8]: Thanks, I had no idea.  Done in 3fcfff5c2d.

File: src/optimizer/rule/date_trunc_simplification.cpp (hunk 9 above)
Comment (rcurtin) [reply src/optimizer/rule/date_trunc_simplification.cpp:52]: Ah, you are right!  Fixed in 1ac88b32b0.

File: src/optimizer/rule/date_trunc_simplification.cpp (hunk 10 above)
Comment (rcurtin) [reply src/optimizer/rule/date_trunc_simplification.cpp]: Ah, way cleaner; done in 2a5ec0ce6f.

File: src/optimizer/rule/date_trunc_simplification.cpp (hunk 11 above)
Comment (rcurtin) [reply src/optimizer/rule/date_trunc_simplification.cpp]: I thought about it when I implemented it and at the time I thought that it was cleaner for a reader to keep the optimizations separate, but that was a weak preference.  I merged them together in d95efbb68d.

File: src/optimizer/rule/date_trunc_simplification.cpp (hunk 12 above)
Comment (rcurtin) [reply src/optimizer/rule/date_trunc_simplification.cpp]: Also done in d95efbb68d.

File: src/optimizer/rule/date_trunc_simplification.cpp (hunk 13 above)
Comment (rcurtin) [reply src/optimizer/rule/date_trunc_simplification.cpp]: Ah, thank you!  I fixed this and added tests for all interval types in 3ec53d7122.

File: test/optimizer/date_trunc_simplification_icu.test (hunk 7 above)
Comment (hawkfish) [reply test/optimizer/date_trunc_simplification_icu.test:14]: No, I mean you should use a different set of [time binning rules](https://duckdb.org/2022/01/06/time-zones.html) than UTC. So `Set TimeZone='Europe/Amsterdam';` for example and then perform the operation across a daylight savings time boundary.

(And no one is a ["timezone expert"](https://xkcd.com/2867/)!)

File: src/optimizer/rule/date_trunc_simplification.cpp (hunk 11 above)
Comment (hawkfish) [reply src/optimizer/rule/date_trunc_simplification.cpp]: Thanks - it's more about not repeating oneself and keeping the code a bit smaller (we've started having problems building the WASM version).

File: test/optimizer/date_trunc_simplification_icu.test (hunk 7 above)
Comment (hawkfish) [reply test/optimizer/date_trunc_simplification_icu.test:14]: `America/St_Johns` should cover both DST and fractional offsets.

Comment (rcurtin) [reply test/optimizer/date_trunc_simplification_icu.test:14]: Ok---I think I understand what you were looking for here and pushed 3bd520dd92 so that the tests run in the St. Johns timezone instead of UTC.  Let me know what you think, or if I missed the goal. :)

**Review by hawkfish — APPROVED** [review]
Thanks for sticking with it!

**Comment by rcurtin** [comment]
Of course!  And thanks for the review.  Happy to help DuckDB get better :)

**Comment by hannes** [comment]
Thanks!
